### PR TITLE
[RFC] StableHLO v1.0 Opset Deprecations & Cleanups

### PR DESCRIPTION
**Pull request by GleasonK**
A proposal to remove redundant operations from StableHLO before long-term compatibility guarantees go into place.

High level summary:
- Remove `CreateTokenOp`, `TraceOp`, `BroadcastOp`, `DotOp`, `UnaryEinsumOp`, `RealDynamicSliceOp`.
- Enhance `DynamicSliceOp`.
- Move `CrossReplicaSumOp` to CHLO.
- Hopefully remove/move to CHLO (need feedback) `MapOp`, `RngOp`, `EinsumOp`, `TorchIndexSelectOp`, `GetTupleElementOp`, and `tuple` type.

OpenXLA Discuss post: https://groups.google.com/a/openxla.org/g/openxla-discuss/c/sBAkvnd2bcA

Related tickets: #2176, #3

### CODE REVIEW

File: rfcs/20240503-opset-deprecations.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+Author: gleasonk<br/>
+Last Modified: 5/3/24<br/>
+Status: In review<br/>
+
+## Background
+
+This doc covers a list of opset cleanups that we want to do for StableHLO v1.0.
+Most of these ops were never spec’ed and therefore have no formal compatibility
+guarantees, per the Unspecced Features compatibility exemption, however we can
```
Comment (mlevesquedion) [inline rfcs/20240503-opset-deprecations.md:11]: Add a link? https://github.com/openxla/stablehlo/blob/main/docs/compatibility.md#out-of-scope

File: rfcs/20240503-opset-deprecations.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+with [`einsum` lowering][einsum-lowering].
+These ops will follow the formal deprecation process listed above.
+
+Helper methods can be added to the support op for compatibility, something like:
+`isLhsBroadcast`, `isSimpleDot`, `isUnaryEinsum`.
+
+### P2: Deprecate `RealDynamicSliceOp`, Enhance `DynamicSliceOp`
+
+In terms of naming `stablehlo.dynamic_slice` is more in-model than
+`real_final_v2_dynamic_slice`. However in terms of functionality, per
```
Comment (mlevesquedion) [inline rfcs/20240503-opset-deprecations.md:54]: Using the actual name of the op may be clearer, though I do think this is amusing.

File: rfcs/20240503-opset-deprecations.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+term since frameworks map to this op, and this will keep the refactoring fairly
+trivial.
+
+### P4: (Feedback Requested) Deprecate `MapOp`, `RngOp`, `EinsumOp`, `TorchIndexSelectOp`
+
+These are all ops that seem to have very limited use in StableHLO. It would be
+great to remove them all or move them to CHLO, as opposed to providing long term
+compatibility on ops that aren't needed.
+
+In the interim, we only plan to guarantee the existing 6mo compatibility
```
Comment (mlevesquedion) [inline rfcs/20240503-opset-deprecations.md:80]: Nit: consider spelling out 6mo. I don't think 6mo is particularly unclear, but 6 months is clearer.

File: rfcs/20240503-opset-deprecations.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+guarantees until these ops until their futures are more clearly known.
```
Comment (mlevesquedion) [inline rfcs/20240503-opset-deprecations.md:81]:
```suggestion
guarantees until these ops' futures are more clearly known.
```

File: rfcs/20240503-opset-deprecations.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+- **MapOp** is unused as far as we can tell, including in HLO. It’s uses tend to
```
Comment (mlevesquedion) [inline rfcs/20240503-opset-deprecations.md:83]:
```suggestion
- **MapOp** is unused as far as we can tell, including in HLO. Its uses tend to
```

File: rfcs/20240503-opset-deprecations.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+be just for a region to mimic a composite, which is no longer needed after the
+addition of the `stablehlo.composite` op. This op likely can be removed.
+- **RngOp** is stateful, and there is a better alternative in RngBitGeneratorOp.
+- **EinsumOp** can likely be moved to CHLO, the [xla::Einsum][einsum] method is
+similarly a decomposition. It is unclear how necessary this abstraction is for
+linalg lowerings though.
```
Comment (mlevesquedion) [inline rfcs/20240503-opset-deprecations.md:89]: Consider being more explicit, e.g. "this abstraction does not seem necessary for linalg lowerings."

**Review by mlevesquedion — APPROVED** [review]

File: rfcs/20240503-opset-deprecations.md (hunk 6 above)
Comment (GleasonK) [reply rfcs/20240503-opset-deprecations.md:89]: It's not clear if it is required or not, need to do a bit more digging / asking around before I make that statement :)

File: rfcs/20240503-opset-deprecations.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
+# RFC: StableHLO v1.0 Opset Deprecations & Cleanups
+
+Author: gleasonk<br/>
+Last Modified: 5/3/24<br/>
+Status: In review<br/>
```
Comment (GleasonK) [inline rfcs/20240503-opset-deprecations.md:5]: Opening blocking comment to avoid accidental merge.

**Review by sdasgup3 — APPROVED** [review]

File: docs/spec.md [hunk 8]
```diff
@@ -6559,6 +6558,68 @@ tensor. Depending on the element type, does the following:
...
+  the StableHLO opset but have been later deemed to not fit it well:
+  `broadcast`, `create_token`, `cross-replica-sum`, `dot`, `einsum`,
+  `torch_index_select`, `unary_einsum`
+  ([#3](https://github.com/openxla/stablehlo/issues/3)).
+* Unused ops - These operations may have been useful at some point, but the ops
+  were either underdeveloped, or the pipelines using these ops have been
+  refactored to not require them anymore. This includes `map`, `tuple` and
+  `get_tuple_element`, `rng`.
+
+Some of these ops can confidently be easily removed (`broadcast`,
```
Comment (mlevesquedion) [inline docs/spec.md:6615]: Nit: "can confidently be easily removed" sounds a little weird -- consider rephrasing and maybe providing additional context: "can be removed easily given that they can be expressed using existing ops"

**Review by mlevesquedion — APPROVED** [review]

File: docs/spec.md [hunk 9]
```diff
@@ -6559,6 +6558,68 @@ tensor. Depending on the element type, does the following:
...
+
+There are several StableHLO operations that were inherited from MHLO which are
+deprecated and on the way out of StableHLO. The full details on these removals
+can be found in the [StableHLO v1.0 Cleanup #2283](https://github.com/openxla/stablehlo/pull/2283).
+
+These operations fall into a few categories:
+
+* "Not in HLO" category of StableHLO operations - they were initially part of
+  the StableHLO opset but have been later deemed to not fit it well:
+  `broadcast`, `create_token`, `cross-replica-sum`, `dot`, `einsum`,
```
[the remark this answers is not in the document]
Comment (GleasonK) [reply docs/spec.md:6607]: Nope.. it's an awfully named op..

**Pull request merged**